Update torch::stable::Tensor() default constructor

mikaylagawarecki opened the pull request:
Allows things like

```cpp
Tensor cu_seqlens_q;
if (...) {
   cu_seqlens_q = ...
} 
...
```

Also adds `torch::stable::Tensor.defined()`

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #159328
* __->__ #159507

mikaylagawarecki committed:
Update torch::stable::Tensor() default constructor

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Update torch::stable::Tensor() default constructor"

Allows things like

```cpp
Tensor cu_seqlens_q;
if (...) {
   cu_seqlens_q = ...
} 
...
```




[ghstack-poisoned]

janeyx99 commented on torch/csrc/stable/tensor.h:
```diff
@@ -31,7 +31,13 @@ class Tensor {
 
  public:
-  Tensor() = delete;
+  Tensor() {
```
Document what this does -- does it create new memory? Who is responsible for cleaning up the memory? (I expect the answer is that this object will clean it up properly). Along that thread, let's also ensure we have tests for memory with this constructor.

janeyx99 commented on torch/csrc/stable/tensor.h:
```diff
@@ -118,6 +124,12 @@ class Tensor {
     return size;
   }
 
+  bool defined() const {
```
This is the same semantic as the one in TensorBase.h right?

And to make sure I'm understanding correctly: an undefined tensor is an uninitialized tensor which I know has no memory but is it basically a nullptr?

mikaylagawarecki replied:
Yes and yes

My understanding is `TensorBase() = default` constructor would initialize its `c10::intrusive_ptr<TensorImpl, UndefinedTensorImpl> impl_` to a nullptr indeed

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/ops.py:
```diff
@@ -164,3 +164,15 @@ def fill_infinity(t) -> Tensor:
     return torch.ops.libtorch_agnostic.fill_infinity.default(t)
+
+
+def test_default_constructor(undefined) -> bool:
```
A bit confusing that this takes in "undefined" but the kernel.cpp version takes in defined. We should converge on defined?

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/ops.py:
```diff
@@ -164,3 +164,15 @@ def fill_infinity(t) -> Tensor:
+    Tests the default constructor for torch::stable::Tensor.
+
+    Args:
+        undefined: bool - if True, tests defined tensor; if False, tests undefined tensor
```
You basically already have undefined semantically meaning defined 😛

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -218,6 +218,16 @@ def test_fill_infinity(self, device):
+        def test_default_constructor(self):
+            import libtorch_agnostic
+
+            is_defined_defined = libtorch_agnostic.ops.test_default_constructor(True)
+            self.assertTrue(is_defined_defined)
```
```suggestion
            defined_tensor_is_defined = libtorch_agnostic.ops.test_default_constructor(True)
            self.assertTrue(defined_tensor_is_defined)
```
nit for readability

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -218,6 +218,16 @@ def test_fill_infinity(self, device):
+            is_defined_defined = libtorch_agnostic.ops.test_default_constructor(True)
+            self.assertTrue(is_defined_defined)
+
+            is_undefined_defined = libtorch_agnostic.ops.test_default_constructor(False)
+            self.assertFalse(is_undefined_defined)
```
```suggestion
            undefined_tensor_is_defined = libtorch_agnostic.ops.test_default_constructor(False)
            self.assertFalse(undefined_tensor_is_defined)
```

janeyx99 reviewed (commented):
Approach looks good to me! Let's add memory tests

mikaylagawarecki committed:
Update on "Update torch::stable::Tensor() default constructor"

Allows things like

```cpp
Tensor cu_seqlens_q;
if (...) {
   cu_seqlens_q = ...
} 
...
```

Also adds `torch::stable::Tensor.defined()`




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -222,11 +222,52 @@ def test_fill_infinity(self, device):
+                    libtorch_agnostic.ops.test_default_constructor(False)
+
+                final_memory = process.memory_info().rss
+                memory_increase = final_memory - initial_memory
```
woah a way to test CPU memory! how confident are you that this test is testing the right thing? This looks okay to me but I am not an expert

mikaylagawarecki replied:
I tried creating a list of tensors an the RSS increased, so I felt somewhat confident this was doing the right thing

cc @albanD do you know :P

albanD replied:
This will most likely be flaky as we don't control what else might grab cpu memory.
I would recommend running this test on cuda and check for cuda memory which is very reliable for this kind of things.
Don't forget to add @serialTest() decorator to avoid cross polution of course ;)

mikaylagawarecki committed:
Update on "Update torch::stable::Tensor() default constructor"

Allows things like

```cpp
Tensor cu_seqlens_q;
if (...) {
   cu_seqlens_q = ...
} 
...
```

Also adds `torch::stable::Tensor.defined()`




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Update torch::stable::Tensor() default constructor"

Allows things like

```cpp
Tensor cu_seqlens_q;
if (...) {
   cu_seqlens_q = ...
} 
...
```

Also adds `torch::stable::Tensor.defined()`




[ghstack-poisoned]

mikaylagawarecki committed:
Update

[ghstack-poisoned]
